aten/src/ATen/core/List_inl.h

```diff
@@ -43,7 +43,8 @@ List<T>::List(TypePtr elementType)
 : List(make_intrusive<c10::detail::ListImpl>(
     typename c10::detail::ListImpl::list_type(),
     std::move(elementType))) {
-  static_assert(std::is_same<T, IValue>::value, "This constructor is only valid for c10::impl::GenericList.");
+  static_assert(std::is_same<T, IValue>::value || std::is_same<T, c10::intrusive_ptr<ivalue::Future>>::value,
```

Comment:
@wanchaol is this change OK?

Reply:
Oh, and to clarify this one - the issue is that `ivalue::Future` doesn't (and maybe can't trivially) define a `getTypePtr_::call()` operator to get its type, since the type isn't part of the template.

Without  an exception to the assert, it's not possible to have a `c10::List` that stores objects that don't have the `call()` operator defined.

Reply:
@wanchaol Could you take a look, particularly this line?

Reply:
hmmm sorry for the late check. I am wondering why we need this check, in particular, `GenericList` already stores a list of ivalues, including the possible ivalue::Future (of course you can't directly store c10::List<ivalue::Future>, but we can use c10::List<IValue> already, where the IValue can be turned into ivalue::Future by `ivalue.toFuture`.

Reply:
Yes, it is possible to instantiate a `GenericList` of generic `IValues` (though I think the api is technically internal only?)

For this api, I was hoping to make the api slightly more typed, i.e. `List<Future>`

That said, without this one-liner, it's simply not possible to instantiate a `List<Future>` (as it is possible to instantiate `List<Tensor>`, `List<int64_t>`, etc) without a compile error. This part seems kind of odd from an api consistency standpoint.

Here's a concrete code snippet:
```
   // Future to insert
    auto f1 = make_intrusive<Future>(IntType::get());                                                                                    
    // Untyped list works just fine
    c10::impl::GenericList genericList(AnyType::get());                                                                                       
    genericList.push_back(f1);                                                                                                                

    // This will not compile without the 1-liner above, because of the lack of a 
   // definition (or good way of doing so) of the "call()" operator for Futures
    List<intrusive_ptr<ivalue::Future>> typedList(FutureType::create(AnyType::get()));                                                   
    typedList.push_back(f1);                                                                                                                  
```

I think it would be nice to be able to create List<Future> on occasion when it's desired... that's all that this line attempts to do.